## Практическое задание №2 «Создание признакового пространства»
Продолжим обработку данных с Твиттера.

In [1]:
import columns as columns
import pandas as pd

In [4]:
df = pd.read_pickle('tweet_correction.pkl')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49159 entries, 0 to 49158
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    49159 non-null  int64  
 1   label                 31962 non-null  float64
 2   tweet                 49159 non-null  object 
 3   cor_tweet             49159 non-null  object 
 4   tweet_token           49159 non-null  object 
 5   tweet_token_filtered  49159 non-null  object 
 6   tweet_stemmed         49159 non-null  object 
 7   tweet_lemmatized      49159 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 3.0+ MB


### 1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform().
Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с
помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features=1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с
помощью CountVectorizer.get_feature_names().

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd


def dummy(doc):
    return doc


count_vectorizer = CountVectorizer(binary=False,
                                   analyzer='word',
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english',
                                   tokenizer=dummy,
                                   preprocessor=dummy,
                                   token_pattern=None)


In [9]:
bag_of_words_stemmed = count_vectorizer.fit_transform(df.tweet_stemmed)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns=feature_names).sample(5)

C:\Users\k142\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
39270    0        0       0        0    0       0      0       0   0      0   
41317    0        0       0        0    0       0      0       0   0      0   
45818    0        0       0        0    0       0      0       0   0      0   
11960    0        0       0        0    0       0      0       0   0      0   
29607    0        0       0        0    0       0      0       0   0      0   

       ...  yeah  year  yesterday  yo  yoga  york  young  youtub  yr  yummi  
39270  ...     0     0          0   0     0     0      0       0   0      0  
41317  ...     0     0          0   0     0     0      0       0   0      0  
45818  ...     0     0          0   0     0     0      0       0   0      0  
11960  ...     0     0          0   0     0     0      0       0   0      0  
29607  ...     0     0          0   0     0     0      0       0   0      0  

[5 rows x 1000 columns]

In [10]:
bag_of_words_lemmatized = count_vectorizer.fit_transform(df.tweet_lemmatized)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns=feature_names).sample(5)

C:\Users\k142\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


able  absolutely  account  act  action  actor  actually  adapt  add  \
26031     0           0        0    0       0      0         0      0    0   
20683     0           0        0    0       0      0         0      0    0   
38632     0           0        0    0       0      0         0      0    0   
19802     0           0        0    0       0      0         0      0    0   
25407     0           0        0    0       0      0         0      0    0   

       adventure  ...  year  yes  yesterday  yo  yoga  york  young  youtube  \
26031          0  ...     0    0          0   0     0     0      0        0   
20683          0  ...     0    0          0   0     0     0      0        0   
38632          0  ...     0    0          0   0     0     0      0        0   
19802          0  ...     0    0          0   0     0     0      0        0   
25407          0  ...     0    0          0   0     0     0      0        0   

       yr  yummy  
26031   0      0  
20683   0      0  
38632   0      0  
19802   0      0  
25407   0      0  

[5 rows x 1000 columns]

### 2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform().
Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с
помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features=1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с
помощью TfidfVectorizer.get_feature_names().

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer, HashingVectorizer


def dummy(doc):
    return doc


tfidf_vectorizer = TfidfVectorizer(binary=False,
                                   analyzer='word',
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english',
                                   tokenizer=dummy,
                                   preprocessor=dummy,
                                   token_pattern=None)

In [15]:
bag_of_words_stemmed = tfidf_vectorizer.fit_transform(df.tweet_stemmed)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names_out()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns=feature_names).sample(5)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
38176  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
45330  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
43582  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
38259  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
21345  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  yeah  year  yesterday   yo  yoga  york  young  youtub   yr  yummi  
38176  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
45330  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
43582  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
38259  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
21345  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  

[5 rows x 1000 columns]

In [14]:
bag_of_words_lemmatized = tfidf_vectorizer.fit_transform(df.tweet_lemmatized)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names_out()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns=feature_names).sample(5)

able  absolutely  account  act  action  actor  actually  adapt  add  \
7027    0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
28610   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
19770   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
16690   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
40302   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   

       adventure  ...  year  yes  yesterday   yo  yoga  york  young  youtube  \
7027         0.0  ...   0.0  0.0        0.0  0.0   0.0   0.0    0.0      0.0   
28610        0.0  ...   0.0  0.0        0.0  0.0   0.0   0.0    0.0      0.0   
19770        0.0  ...   0.0  0.0        0.0  0.0   0.0   0.0    0.0      0.0   
16690        0.0  ...   0.0  0.0        0.0  0.0   0.0   0.0    0.0      0.0   
40302        0.0  ...   0.0  0.0        0.0  0.0   0.0   0.0    0.0      0.0   

        yr  yummy  
7027   0.0    0.0  
28610  0.0    0.0  
19770  0.0    0.0  
16690  0.0    0.0  
40302  0.0    0.0  

[5 rows x 1000 columns]


### 3. Проверьте ваши векторайзеры на корпусе который использовали на вебинаре,

составьте таблицу метод векторизации и скор который вы получили (в методах векторизации по
изменяйте параметры что бы добиться лучшего скора) обратите внимание как
падает/растёт скор при уменьшении количества фичей, и изменении параметров, так же
попробуйте применить к векторайзерам PCA для сокращения размерности посмотрите на
качество сделайте выводы

In [16]:
# Загружаем данные
data = open('corpus').read()
labels, texts = [], []
for i, line in enumerate(data.split("\n")):
    content = line.split()
    labels.append(content[0])
    texts.append(" ".join(content[1:]))

# создаем df
trainDF = pd.DataFrame()
trainDF['text'] = texts
trainDF['label'] = labels
trainDF.head(5)

text       label
0  Stuning even for the non-gamer: This sound tra...  __label__2
1  The best soundtrack ever to anything.: I'm rea...  __label__2
2  Amazing!: This soundtrack is my favorite music...  __label__2
3  Excellent Soundtrack: I truly like this soundt...  __label__2
4  Remember, Pull Your Jaw Off The Floor After He...  __label__2

In [17]:
from sklearn import model_selection, preprocessing, linear_model
from sklearn.metrics import accuracy_score

train_x, valid_x, train_y, valid_y = model_selection.train_test_split(trainDF['text'], trainDF['label'])

# labelEncode целевую переменную
encoder = preprocessing.LabelEncoder()
train_y = encoder.fit_transform(train_y)
valid_y = encoder.fit_transform(valid_y)

In [37]:
stat_test_vectorizes = []

In [40]:
def test_vectorizes(vectorizer, classifier, corpus, train_x, valid_x, train_y, valid_y):
    vectorizer.fit(corpus)
    xtrain_count = vectorizer.transform(train_x)
    xvalid_count = vectorizer.transform(valid_x)

    classifier.fit(xtrain_count, train_y)
    predictions = classifier.predict(xvalid_count)
    print('размерность вектора', xtrain_count.shape)

    return accuracy_score(valid_y, predictions)

> Бейслан простой эмбеддинг на основе частотности

In [41]:
count_vect = CountVectorizer(analyzer='word', token_pattern=r'\w{1,}', binary=False)
classifier = linear_model.LogisticRegression()
score = test_vectorizes(count_vect, classifier, trainDF['text'], train_x, valid_x, train_y, valid_y)
stat_test_vectorizes.append(['CountVectorizer', score])

размерность вектора (7500, 31681)


C:\Users\k142\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [43]:
count_vect = HashingVectorizer(analyzer='word', token_pattern=r'\w{1,}', n_features=3000)
classifier = linear_model.LogisticRegression(max_iter=300)
score = test_vectorizes(count_vect, classifier, trainDF['text'], train_x, valid_x, train_y, valid_y)
stat_test_vectorizes.append(['HashingVectorizer3000', score])

размерность вектора (7500, 3000)


In [44]:
count_vect = HashingVectorizer(analyzer='word', token_pattern=r'\w{1,}', n_features=100)
classifier = linear_model.LogisticRegression(max_iter=300)
score = test_vectorizes(count_vect, classifier, trainDF['text'], train_x, valid_x, train_y, valid_y)
stat_test_vectorizes.append(['HashingVectorizer1000', score])

размерность вектора (7500, 100)


In [45]:
count_vect = TfidfVectorizer(analyzer='word', token_pattern=r'\w{1,}', max_features=3000)
classifier = linear_model.LogisticRegression(max_iter=300)
score = test_vectorizes(count_vect, classifier, trainDF['text'], train_x, valid_x, train_y, valid_y)
stat_test_vectorizes.append(['TfidfVectorizer3000', score])

размерность вектора (7500, 3000)


In [46]:
count_vect = TfidfVectorizer(analyzer='word', token_pattern=r'\w{1,}', max_features=300)
classifier = linear_model.LogisticRegression(max_iter=300)
score = test_vectorizes(count_vect, classifier, trainDF['text'], train_x, valid_x, train_y, valid_y)
stat_test_vectorizes.append(['TfidfVectorizer300', score])

размерность вектора (7500, 300)


In [52]:
pd.DataFrame(stat_test_vectorizes, columns=['Векторайзер', 'Точность'])

Векторайзер  Точность
0        CountVectorizer    0.8568
1  HashingVectorizer3000    0.8308
2  HashingVectorizer1000    0.6716
3    TfidfVectorizer3000    0.8700
4     TfidfVectorizer300    0.8248

#### Вывод:

Корпус данных включает 7,5 тысячи документов и словарь в 35 тысяч токенов. Задача заключалась в построении вектора (простейшего эмбеддинга) который сможет уловить закономерности в документах. Проверка способности уловить закономерности проверялась на решении задачи классификации логистической регрессии. Простейший частотный словарь дал приличную оценку - 86%
Лучший результат показал механизм TfIdf при этом словарь был снижен в 10 раз. И даже при снижении словаря на два порядка (в 100 раз) точность осталась на уровне 83%.